src/audio/mfcc/mfcc.c

```diff
@@ -0,0 +1,282 @@
+{
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+	struct audio_stream __sparse_cache *source = input_buffers->data;
+	struct audio_stream __sparse_cache *sink = output_buffers->data;
+	int frames = input_buffers->size;
+
+	comp_dbg(mod->dev, "mfcc_process(), start");
+
+	frames = MIN(frames, cd->max_frames);
+	cd->mfcc_func(mod, input_buffers, output_buffers, frames);
+
+	/* TODO: use module_update_buffer_position() from #6194 */
+	input_buffers->consumed += audio_stream_frame_bytes(source) * frames;
+	output_buffers->size += audio_stream_frame_bytes(sink) * frames;
+	comp_dbg(mod->dev, "mfcc_process(), done");
+	return 0;
+}
+
+static void mfcc_set_alignment(struct audio_stream *source, struct audio_stream *sink)
+{
+	const uint32_t byte_align = 1;
+	const uint32_t frame_align_req = 1;
+
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, source);
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, sink);
+}
+
+static int mfcc_prepare(struct processing_module *mod)
+{
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	struct comp_buffer __sparse_cache *source_c;
+	struct comp_buffer __sparse_cache *sink_c;
+	struct comp_dev *dev = mod->dev;
+	enum sof_ipc_frame source_format;
+	enum sof_ipc_frame sink_format;
+	uint32_t sink_period_bytes;
+	int ret;
+
+	comp_info(dev, "mfcc_prepare()");
+
+	/* MFCC component will only ever have 1 source and 1 sink buffer */
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	source_c = buffer_acquire(sourceb);
+	sink_c = buffer_acquire(sinkb);
+
+	/* get source data format */
+	source_format = source_c->stream.frame_fmt;
+
+	/* set align requirements */
+	mfcc_set_alignment(&source_c->stream, &sink_c->stream);
+
+	/* get sink data format and period bytes */
+	sink_format = sink_c->stream.frame_fmt;
+	sink_period_bytes = audio_stream_period_bytes(&sink_c->stream, dev->frames);
+	comp_info(dev, "mfcc_prepare(), source_format = %d, sink_format = %d",
+		  source_format, sink_format);
+	if (sink_c->stream.size < sink_period_bytes) {
+		comp_err(dev, "mfcc_prepare(): sink buffer size %d is insufficient < %d",
+			 sink_c->stream.size, sink_period_bytes);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
+
+	/* Initialize MFCC, max_frames is set to dev->frames + 4 */
+	if (cd->config) {
+		ret = mfcc_setup(mod, dev->frames + 4, source_c->stream.rate,
+				 source_c->stream.channels);
+		if (ret < 0) {
+			comp_err(dev, "mfcc_prepare(), setup failed.");
+			goto err;
+		}
+	}
+
+	cd->mfcc_func = mfcc_find_func(source_format, sink_format, mfcc_fm, ARRAY_SIZE(mfcc_fm));
+	if (!cd->mfcc_func) {
+		comp_err(dev, "mfcc_prepare(), No proc func");
+		ret = -EINVAL;
+		goto err;
+	}
+
+	buffer_release(source_c);
+	buffer_release(sink_c);
```

Comment:
the release order must be symmetric to acquisition: AB-BA, think of them as of any locking - mutexes, spinlocks,... these two have to be swapped

Reply:
@singalsu this one not optional. Needs to be fixed.

Reply:
I didn't know, thanks for highlighting this. I suppose this is beyond scope of valgrind. Wish there would be a tool that could detect wrong order. E.g. testbench version of buffer acquire and release that asserts when not "stack" order.

Reply:
This and next one are now fixed